README.md
## uxcore-timeline

React timeline

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test Coverage][coveralls-image]][coveralls-url]
[![Dependency Status][dep-image]][dep-url]
[![devDependency Status][devdep-image]][devdep-url] 
[![NPM downloads][downloads-image]][npm-url]

[![Sauce Test Status][sauce-image]][sauce-url]

[npm-image]: http://img.shields.io/npm/v/uxcore-timeline.svg?style=flat-square
[npm-url]: http://npmjs.org/package/uxcore-timeline
[travis-image]: https://img.shields.io/travis/uxcore/uxcore-timeline.svg?style=flat-square
[travis-url]: https://travis-ci.org/uxcore/uxcore-timeline
[coveralls-image]: https://img.shields.io/coveralls/uxcore/uxcore-timeline.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/uxcore/uxcore-timeline?branch=master
[dep-image]: http://img.shields.io/david/uxcore/uxcore-timeline.svg?style=flat-square
[dep-url]: https://david-dm.org/uxcore/uxcore-timeline
[devdep-image]: http://img.shields.io/david/dev/uxcore/uxcore-timeline.svg?style=flat-square
[devdep-url]: https://david-dm.org/uxcore/uxcore-timeline#info=devDependencies
[downloads-image]: https://img.shields.io/npm/dm/uxcore-timeline.svg
[sauce-image]: https://saucelabs.com/browser-matrix/uxcore-timeline.svg
[sauce-url]: https://saucelabs.com/u/uxcore-timeline


### Development

```sh
git clone https://github.com/uxcore/uxcore-timeline
cd uxcore-timeline
npm install
npm run server
```

if you'd like to save your install time，you can use uxcore-tools globally.

```sh
npm install uxcore-tools -g
git clone https://github.com/uxcore/uxcore-timeline
cd uxcore-timeline
npm install
npm run dep
npm run start
```

### Test Case

```sh
npm run test
```

### Coverage

```sh
npm run coverage
```

## Demo

http://uxcore.github.io/components/timeline

## Contribute

Yes please! See the [CONTRIBUTING](https://github.com/uxcore/uxcore/blob/master/CONTRIBUTING.md) for details.

## 何时使用

- 当有一系列信息需要从上至下按时间排列时；
- 需要有一条时间轴进行视觉上的串联时；

## API

```jsx
<Timeline>
    <Timeline.Item title='2017-10-25'>
        <p>content1</p>
        <p>content1</p>
    </Timeline.Item>
    <Timeline.Item title='2017-10-26' active>
        <p>content2</p>
        <p>content2</p>
    </Timeline.Item>
    <Timeline.Item title='2017-10-27'>
        <p>content3</p>
        <p>content3</p>
    </Timeline.Item>
</Timeline>
```

## Props

### Timeline

时间轴。

| Name | Type | Required | Default | Comments |
|---|---|---|---|---|
| className | String | false | 无 | 额外类名 |
| pending | jsx | false | 无 | 指定最后一个幽灵节点内容 |

### Timeline.Item

时间轴的每一个节点。

| Name | Type | Required | Default | Comments |
|---|---|---|---|---|
| className | String | false | 无 | 额外类名 |
| color | String | false | 无 | 指定节点的色调 `orange, blue, gray, green`，或自定义的色值 |
| dot | jsx | false | 无 | 自定义时间轴点 |
| dotted | Boolean | false | false | 虚线连接线 |
| active | Boolean | false | false | 指定当前时间节点or处理中状态 |
| title | String | false | 无 | 节点的标题 |


HISTORY.md
# history

## 0.2.6

* `CHANGED` support js style export

## 0.2.4

* `FIXED` React 16.x compatible

## 0.2.3

* `CHANGED` UI Upgrade

## 0.2.1

* `FIXED` React 15.x compatible

## 0.2.0

* `NEW` add new props `dotted`, `active` & `title`
* `CHANGED` new style


tests/Timeline.spec.js
import expect from 'expect.js';
import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Icon from 'uxcore-icon';
import Timeline from '../src';

Enzyme.configure({ adapter: new Adapter() });

describe('Timeline', () => {
  it('should render correctly', () => {
    mount(<Timeline />);
  });
  it('props', () => {
    const wrapper = mount(<Timeline pending={<a>To do</a>}>
      <Timeline.Item title="2018-10-25 星期一" dotted>
        <p>content1</p>
        <p>content1</p>
      </Timeline.Item>
      <Timeline.Item title="2018-10-26 星期二" dot={<Icon name="shezhi" />}>
        <p>content2</p>
        <p>content2</p>
      </Timeline.Item>
      <Timeline.Item title="2018-10-27 星期三">
        <p>content3</p>
        <p>content3</p>
      </Timeline.Item>
    </Timeline>);
    expect(wrapper.find('.uxicon-shezhi')).to.have.length(1);
    expect(wrapper.find('.kuma-timeline-item-title')).to.have.length(3);
    expect(wrapper.find('.kuma-timeline-item-dotted')).to.have.length(1);
    expect(wrapper.find(Timeline)).to.have.length(1);
    expect(wrapper.contains(<a>To do</a>)).to.equal(true);
  });
});


src/Timeline.js
/**
 * Timeline Component for uxcore
 * @author muwen.lb
 *
 * Copyright 2015-2016, Uxcore Team, Alinw.
 * All rights reserved.
 */
const React = require('react');
const TimelineItem = require('./TimelineItem');

class Timeline extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      pending: props.pending,
    };
  }

  render() {
    const me = this;
    let pending = me.state.pending;
    return (
      <div className={'kuma-timeline ' + me.props.className}>
        <ul className={!!pending ? 'kuma-timeline-pending' : ''}>
          {
            React.Children.map(me.props.children, (ele, idx) => {
              return React.cloneElement(ele, {
                last: idx === me.props.children.length - 1,
              });
            })
          }
          {(!!pending) ? <TimelineItem pending={true}>{pending}</TimelineItem> : null}
        </ul>
      </div>
    );
  }
}

// http://facebook.github.io/react/docs/reusable-components.html
Timeline.defaultProps = {
  className: '', //额外的className
};

Timeline.propTypes = {
  className: React.PropTypes.string,
};

Timeline.Item = TimelineItem;

Timeline.displayName = 'Timeline';

module.exports = Timeline;


src/TimelineItem.js
/**
 * Timeline Component for uxcore
 * @author muwen.lb
 *
 * Copyright 2015-2016, Uxcore Team, Alinw.
 * All rights reserved.
 */
const React = require('react');
const classnames = require('classnames');

class TimelineItem extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      dot: props.dot,
    };
  }

  render() {
    const me = this;
    const itemClassName = classnames({
      'kuma-timeline-item': true,
      'kuma-timeline-item-last': me.props.last,
      'kuma-timeline-item-pending': me.props.pending,
      [`kuma-timeline-item-${me.props.color}`]: true,
      [me.props.className]: me.props.className,
    });

    const dotClassName = classnames({
      'kuma-timeline-item-head': true,
      'kuma-timeline-item-head-custom': me.state.dot,
    });

    return (
      <li className={itemClassName}>
        <div className="kuma-timeline-item-tail" style={{ borderColor: /orange|blue|gray|green/.test(me.props.color) ? null : me.props.color }} />
        <div className={dotClassName} >
          <div className="kuma-timeline-item-head-icon" style={{ backgroundColor: /orange|blue|gray|green/.test(me.props.color) ? null : me.props.color }}>
            {me.state.dot}
          </div>
        </div>
        <div className="kuma-timeline-item-content">
          {me.props.children}
        </div>
      </li>
    );
  }
}

// http://facebook.github.io/react/docs/reusable-components.html
TimelineItem.defaultProps = {
  pending: false, // 'zh-cn'、'en'
  className: '', // 额外的className
  last: false, // 是否最后一个
  color: 'orange', // 圆圈颜色,blue/gray/green
  // dot: false // 自定义圆圈图标
};

TimelineItem.propTypes = {
  pending: React.PropTypes.bool,
  className: React.PropTypes.string,
  last: React.PropTypes.bool,
  color: React.PropTypes.string,
  // dot: React.PropTypes.element
};

TimelineItem.displayName = 'Timeline';

module.exports = TimelineItem;


demo/TimelineDemo.js
/**
 * Timeline Component Demo for uxcore
 * @author muwen.lb
 *
 * Copyright 2015-2016, Uxcore Team, Alinw.
 * All rights reserved.
 */

const React = require('react');
const Timeline = require('../src');

class Demo extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
    };
  }

  render() {
    return (
      <div>
        普通使用：
        <Timeline>
          <Timeline.Item>
            <p>2016-10-25</p>
            <p>content1</p>
            <p>content1</p>
          </Timeline.Item>
          <Timeline.Item>
            <p>2016-10-25</p>
            <p>content1</p>
            <p>content1</p>
          </Timeline.Item>
          <Timeline.Item color="gray">
            <p>2016-10-28</p>
            <p>content4</p>
            <p>content4</p>
          </Timeline.Item>
          <Timeline.Item color="gray">
            <p>2016-10-28</p>
            <p>content4</p>
            <p>content4</p>
          </Timeline.Item>
        </Timeline>
        各种颜色线条（可自定义）：
        <Timeline>
          <Timeline.Item>
            <p>2016-10-25</p>
            <p>content1</p>
            <p>content1</p>
          </Timeline.Item>
          <Timeline.Item color="green">
            <p>2016-10-26</p>
            <p>content2</p>
            <p>content2</p>
          </Timeline.Item>
          <Timeline.Item color="blue">
            <p>2016-10-27</p>
            <p>content3</p>
            <p>content3</p>
          </Timeline.Item>
          <Timeline.Item color="gray">
            <p>2016-10-28</p>
            <p>content4</p>
            <p>content4</p>
          </Timeline.Item>
        </Timeline>
        额外添加item（虚线条）：
        <Timeline pending={<a href="#">To do</a>}>
          <Timeline.Item>
            <p>2016-10-25</p>
            <p>content1</p>
            <p>content1</p>
          </Timeline.Item>
          <Timeline.Item color="green">
            <p>2016-10-26</p>
            <p>content2</p>
            <p>content2</p>
          </Timeline.Item>
          <Timeline.Item color="blue">
            <p>2016-10-27</p>
            <p>content3</p>
            <p>content3</p>
          </Timeline.Item>
          <Timeline.Item color="gray">
            <p>2016-10-28</p>
            <p>content4</p>
            <p>content4</p>
          </Timeline.Item>
        </Timeline>
      </div>
    );
  }
}

module.exports = Demo;
